### PR TITLE
Fix Spark 3.4 build errors [databricks]

### LINKED ISSUE #7432: [BUG][SPARK-41713][SPARK-41726] Spark 3.4 build fails.
**Describe the bug**
Meet two compile errors when building Spark 3.4
```
[ERROR] [Error] /home/remziy/working/rapids/spark-rapids/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3625: type arguments [org.apache.spark.sql.execution.command.CreateDataSourceTableAsSelectCommand] do not conform to method dataWriteCmd's type parameter bounds [INPUT <: org.apache.spark.sql.execution.command.DataWritingCommand]
[ERROR] [Error] /home/remziy/working/rapids/spark-rapids/sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/HiveProviderImpl.scala:43: type arguments [org.apache.spark.sql.hive.execution.OptimizedCreateHiveTableAsSelectCommand] do not conform to method dataWriteCmd's type parameter bounds [INPUT <: org.apache.spark.sql.execution.command.DataWritingCommand]
[ERROR] two errors found
```

**Steps/Code to reproduce bug**
```
mvn package -Dbuildver=340 -DskipTests -s ./jenkins/settings.xml -f ./pom.xml
```

**Expected behavior**

**Environment details (please complete the following information)**
 - Environment location: [Standalone, YARN, Kubernetes, Cloud(specify cloud provider)]
 - Spark configuration settings related to the issue

**Additional context**

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>
Closes #7432.

## Rationales of the Changes
This PR aims to fix the compile errors when building Spark 3.4.
You can find more background from #7432.

## Major changes in the PR
Create `GpuRunnableCommand` and `GpuRunnableCommandExec` to extend the `RunnableCommand` in Spark to do columnar computing.

Move many things (related to CTAS) to shim layers. In Spark 311~33X, those classes / traits / objects are subtypes of `GpuDataWritingCommand`, and in Spark 3.4 they are subtypes of `GpuRunnableCommand`

### CODE REVIEW

**Comment by HaoYang670** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HiveProvider.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@ package com.nvidia.spark.rapids
 
 import org.apache.spark.sql.catalyst.expressions.Expression
```
Comment (res-life) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/HiveProvider.scala:19]: Add 2023 to Copyright

Comment (res-life) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HiveProvider.scala:19]: Also, check other files

Comment (HaoYang670) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HiveProvider.scala:19]: Thank you, I will re check all of them again

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

File: sql-plugin/src/main/332/scala/com/nvidia/spark/rapids/shims/SparkShims.scala [hunk 2]
```diff
@@ -16,4 +16,21 @@
 
 package com.nvidia.spark.rapids.shims
 
-object SparkShimImpl extends Spark331PlusShims with AnsiCastRuleShims
+import com.nvidia.spark.rapids._
+
+import org.apache.spark.sql.execution.command.{CreateDataSourceTableAsSelectCommand, DataWritingCommand, RunnableCommand}
+
+object SparkShimImpl extends Spark331PlusShims with AnsiCastRuleShims {
+  override def getDataWriteCmds: Map[Class[_ <: DataWritingCommand],
```
Comment (res-life) [inline sql-plugin/src/main/332/scala/com/nvidia/spark/rapids/shims/SparkShims.scala:24]: Duplicated block. Extract this block to `311until340-all`

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/shims/GpuCreateDataSourceTableAsSelectCommandShims.scala [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2022, NVIDIA CORPORATION.
```
Comment (res-life) [inline sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/shims/GpuCreateDataSourceTableAsSelectCommandShims.scala:2]: Why move this file?

File: sql-plugin/src/main/311until340-all/scala/com/nvidia/spark/rapids/shims/GpuOptimizedCreateHiveTableAsSelectCommandShims.scala [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
```
Comment (res-life) [inline sql-plugin/src/main/311until340-all/scala/com/nvidia/spark/rapids/shims/GpuOptimizedCreateHiveTableAsSelectCommandShims.scala:1]: Why move this file?

**Comment by res-life** [comment]
Duplicated code, can we simplify them?

diff ./sql-plugin/src/main/311until340-all/scala/com/nvidia/spark/rapids/shims/GpuOptimizedCreateHiveTableAsSelectCommandShims.scala ./sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/GpuOptimizedCreateHiveTableAsSelectCommandShims.scala

diff ./sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/shims/GpuCreateDataSourceTableAsSelectCommandShims.scala ./sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/shims/GpuCreateDataSourceTableAsSelectCommandShims.scala

**Comment by res-life** [comment]
`GpuDataWritingCommandExec.scala`
`GpuRunnableCommandExec.scala`

They are almost the same, create new `GpuCommandExecBase` and `GpuCommand` to remove duplicated code.

File: sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/CreateDataSourceTableAsSelectCommandMetaShim.scala [hunk 5]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (res-life) [inline sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/CreateDataSourceTableAsSelectCommandMetaShim.scala:1]: The file name should be CreateDataSourceTableAsSelectCommandMetaShim**s**, missed the trailing `s`

File: sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/shims/GpuCreateDataSourceTableAsSelectCommandShims.scala (hunk 3 above)
Comment (HaoYang670) [reply sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/shims/GpuCreateDataSourceTableAsSelectCommandShims.scala:2]: The type structure is different for Spark 3.4 and pre versions.
For Spark 311 until 340, `GpuCreateHiveTableAsSelectBase` is a subtype of `GpuDataWritingCommand`, but in Spark 3.4, `GpuCreateHiveTableAsSelectBase` is a subtype of `GpuRunnableCommand`

File: sql-plugin/src/main/311until340-all/scala/com/nvidia/spark/rapids/shims/GpuOptimizedCreateHiveTableAsSelectCommandShims.scala (hunk 4 above)
Comment (HaoYang670) [reply sql-plugin/src/main/311until340-all/scala/com/nvidia/spark/rapids/shims/GpuOptimizedCreateHiveTableAsSelectCommandShims.scala:1]: The type structure is different for Spark 3.4 and pre versions.
For Spark 311 until 340, `GpuCreateHiveTableAsSelectBase` is a subtype of `GpuDataWritingCommand`, but in Spark 3.4, `GpuCreateHiveTableAsSelectBase` is a subtype of `GpuRunnableCommand`

File: sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/CreateDataSourceTableAsSelectCommandMetaShim.scala (hunk 5 above)
Comment (HaoYang670) [reply sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/CreateDataSourceTableAsSelectCommandMetaShim.scala:1]: Thank you! Fixed

File: sql-plugin/src/main/332/scala/com/nvidia/spark/rapids/shims/SparkShims.scala (hunk 2 above)
Comment (HaoYang670) [reply sql-plugin/src/main/332/scala/com/nvidia/spark/rapids/shims/SparkShims.scala:24]: Do you mean adding a new trait `Spark31Xuntil34X`?

**Comment by HaoYang670** [comment]
build

File: sql-plugin/src/main/332/scala/com/nvidia/spark/rapids/shims/SparkShims.scala (hunk 2 above)
Comment (res-life) [reply sql-plugin/src/main/332/scala/com/nvidia/spark/rapids/shims/SparkShims.scala:24]: I mean `311until340-all`, this PR is handing a new Spark change from version 340.
So all the previous versions [311, 340) are able to use this code block, we can extract this to a trait to reuse it.

**Comment by res-life** [comment]
Could we create a following-up PR to do some refactoring?
Did the tests pass for Spark33x, Spark32x, etc? The pre-merge only tests Spark311, so we can locally run the tests for other Spark versions. Or this may cause the nightly builds for other Spark versions to fail.

**Review by res-life — APPROVED** [review]
LGTM

**Comment by res-life** [comment]
Please skip 340 tests, I'm working on them.

**Comment by HaoYang670** [comment]
The last one is a known bug #7516.

**Comment by HaoYang670** [comment]
The others are related to my test env, not the changes in this PR.

**Pull request merged**